### PR TITLE
feat: Implement Export Recorded Data endpoint

### LINKED ISSUE #7: Implement Download (export) endpoint
pre ADR: https://docs.edgexfoundry.org/3.0/design/adr/application/0025-Record-and-Replay/#get_2

Add `RecordedData` API to Data Manager interface and mock
Unit test endpoint using Data Manager mock

### PR DESCRIPTION
**Pull request by marcpfuller**
closes #7

**If your build fails**  due to your commit message not passing the build checks, please review the guidelines here: https://github.com/edgexfoundry/app-record-replay/blob/main/.github/CONTRIBUTING.md

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] I am not introducing a breaking change (if you are, flag in conventional commit message with `BREAKING CHANGE:` describing the break)
- [x] I am not introducing a new dependency (add notes below if you are)
- [x] I have added unit tests for the new feature or bug fix (if not, why?)
- [x] I have fully tested (add details below) this the new feature or bug fix (if not, why?)
- [ ] I have opened a PR for the related docs change (if not, why?)
  <link to docs PR>

## Testing Instructions
Run non-secure EdgeX stack
Run `make build`
Run app service with `-cp -d`
Use postman collecting to test GET to `localhost:59712/api/v3/data`

### CODE REVIEW

**Review by jim-wang-intel — COMMENTED** [review]
LGTM, upon Lenny's apporval

File: internal/controller/controller_test.go [hunk 1]
```diff
@@ -387,7 +388,88 @@ func TestHttpController_CancelReplay(t *testing.T) {
...
+		ExpectedError    error
+	}{
+		{
+			Name:             "Valid - 2 events",
+			ExpectedResponse: &recordedEvents,
+			ExpectedStatus:   http.StatusOK,
+			ExpectedError:    nil,
+		},
+		{
+			Name:             "Valid - no events",
```
[the remark this answers is not in the document]
Comment (marcpfuller) [reply internal/controller/controller_test.go:443]: I was wondering about that, I agree. I think we should return `no content`.

File: internal/controller/controller.go [hunk 2]
```diff
@@ -234,15 +242,56 @@ func (c *httpController) exportRecordedData(writer http.ResponseWriter, request
...
 
-	jsonResponse, err := json.Marshal(recordedData)
-	if err != nil {
+	if reflect.DeepEqual(recordedData, &dtos.RecordedData{}) {
+		writer.WriteHeader(http.StatusNoContent)
+		_, _ = writer.Write([]byte("no recorded data found"))
+		return
+	}
+
+	compression := request.URL.Query().Get("Content-Encoding")
```
Comment (lenny-intel) [inline internal/controller/controller.go:251]: I was thinking the query parameter would be `compression`
i.e.
```
/export?compression=GZIP
```
Why did you pick `Content-Encoding` for the query parameter name?
I think the response should sent `Content-Encoding` in the header like you have done. :-)

File: internal/controller/controller.go [hunk 3]
```diff
@@ -234,15 +242,56 @@ func (c *httpController) exportRecordedData(writer http.ResponseWriter, request
...
+	switch compression {
+	case noCompression:
+		jsonResponse, err := json.Marshal(recordedData)
+		if err != nil {
+			writer.WriteHeader(http.StatusInternalServerError)
+			_, _ = writer.Write([]byte("failed to marshal recorded data"))
+			return
+		}
+		writer.WriteHeader(http.StatusOK)
```
Comment (lenny-intel) [inline internal/controller/controller.go:260]: I think we should be setting `"Content-Type", "application/json"` on all our JSON responses.

File: internal/controller/controller.go [hunk 4]
```diff
@@ -227,8 +235,63 @@ func (c *httpController) replayStatus(writer http.ResponseWriter, request *http.
...
+	case zlibCompression:
+		writer.Header().Set("Content-Encoding", "ZLIB")
+		writer.Header().Set("Content-Type", "application/json")
+		writer.WriteHeader(http.StatusOK)
+		zlibWriter := zlib.NewWriter(writer)
+		defer zlibWriter.Close()
+		err = json.NewEncoder(zlibWriter).Encode(&recordedData)
+		if err != nil {
+			writer.WriteHeader(http.StatusInternalServerError)
+			_, _ = writer.Write([]byte(fmt.Sprintf("%s %s: %s", failedDataCompression, gzipCompression, err)))
```
Comment (lenny-intel) [inline internal/controller/controller.go:272]:
```suggestion
			_, _ = writer.Write([]byte(fmt.Sprintf("%s %s: %s", failedDataCompression, zlibCompression, err)))
```

File: internal/controller/controller.go [hunk 5]
```diff
@@ -227,8 +235,63 @@ func (c *httpController) replayStatus(writer http.ResponseWriter, request *http.
...
+			_, _ = writer.Write([]byte("failed to marshal recorded data"))
+			return
+		}
+		writer.WriteHeader(http.StatusOK)
+		_, _ = writer.Write(jsonResponse)
+
+	case zlibCompression:
+		writer.Header().Set("Content-Encoding", "ZLIB")
+		writer.Header().Set("Content-Type", "application/json")
+		writer.WriteHeader(http.StatusOK)
```
Comment (lenny-intel) [inline internal/controller/controller.go:266]: Move this down like it is in gzip case? or does that one need to be moved up?

**Review by lenny-intel — CHANGES_REQUESTED** [review]

**Review by lenny-intel — APPROVED** [review]
LGTM

**Pull request merged**